// @astrox/connection/lib/cjs/ic/icStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ICStorage = exports._deleteStorage = exports.IDENTITY_PROVIDER_ENDPOINT = exports.IDENTITY_PROVIDER_DEFAULT = exports.KEY_ICSTORAGE_WALLET = exports.KEY_ICSTORAGE_DELEGATION = exports.KEY_ICSTORAGE_KEY = void 0;
exports.KEY_ICSTORAGE_KEY = 'identity';
exports.KEY_ICSTORAGE_DELEGATION = 'delegation';
exports.KEY_ICSTORAGE_WALLET = 'wallet';
exports.IDENTITY_PROVIDER_DEFAULT = 'https://identity.ic0.app';
exports.IDENTITY_PROVIDER_ENDPOINT = '#authorize';
async function _deleteStorage(storage) {
    await storage.remove(exports.KEY_ICSTORAGE_KEY);
    await storage.remove(exports.KEY_ICSTORAGE_DELEGATION);
    await storage.remove(exports.KEY_ICSTORAGE_WALLET);
}
exports._deleteStorage = _deleteStorage;
class ICStorage {
    constructor(prefix = 'astrox-', _localStorage) {
        this.prefix = prefix;
        this._localStorage = _localStorage;
    }
    get(key) {
        return Promise.resolve(this._getICStorage().getItem(this.prefix + key));
    }
    set(key, value) {
        this._getICStorage().setItem(this.prefix + key, value);
        return Promise.resolve();
    }
    remove(key) {
        this._getICStorage().removeItem(this.prefix + key);
        return Promise.resolve();
    }
    _getICStorage() {
        if (this._localStorage) {
            return this._localStorage;
        }
        const ls = typeof window === 'undefined'
            ? typeof global === 'undefined'
                ? typeof self === 'undefined'
                    ? undefined
                    : self.localStorage
                : global.localStorage
            : window.localStorage;
        if (!ls) {
            throw new Error('Could not find local storage.');
        }
        return ls;
    }
}
exports.ICStorage = ICStorage;
