// @connect2ic/core/providers/dist/connect2ic-core-providers.esm.js
import { AstroX } from '../astrox/dist/connect2ic-core-providers-astrox.esm.js';
export { AstroX } from '../astrox/dist/connect2ic-core-providers-astrox.esm.js';
import { InternetIdentity } from '../internet-identity/dist/connect2ic-core-providers-internet-identity.esm.js';
export { InternetIdentity } from '../internet-identity/dist/connect2ic-core-providers-internet-identity.esm.js';
import { PlugWallet } from '../plug-wallet/dist/connect2ic-core-providers-plug-wallet.esm.js';
export { PlugWallet } from '../plug-wallet/dist/connect2ic-core-providers-plug-wallet.esm.js';
import { StoicWallet } from '../stoic-wallet/dist/connect2ic-core-providers-stoic-wallet.esm.js';
export { StoicWallet } from '../stoic-wallet/dist/connect2ic-core-providers-stoic-wallet.esm.js';
import { InfinityWallet } from '../infinity-wallet/dist/connect2ic-core-providers-infinity-wallet.esm.js';
export { InfinityWallet } from '../infinity-wallet/dist/connect2ic-core-providers-infinity-wallet.esm.js';
import { NFID } from '../nfid/dist/connect2ic-core-providers-nfid.esm.js';
export { NFID } from '../nfid/dist/connect2ic-core-providers-nfid.esm.js';
import '../../dist/classPrivateFieldGet-c8fedd89.esm.js';
import '@astrox/connection';
import '@astrox/connection/lib/esm/types';
import '@dfinity/auth-client';
import '@dfinity/agent';
import 'ic-stoic-identity';

const defaultProviders = config => {
  return [new AstroX(config), // EarthWallet,
  new InfinityWallet(config), new InternetIdentity(config), new NFID(config), new PlugWallet(config), new StoicWallet(config)];
};
const walletProviders = config => {
  return [new AstroX(config), new PlugWallet(config)];
};

export { defaultProviders, walletProviders };
