// @astrox/connection/lib/cjs/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionMessageKind = exports.SignerMessageKind = exports.PermissionsType = void 0;
var PermissionsType;
(function (PermissionsType) {
    PermissionsType["identity"] = "permissions-identity";
    PermissionsType["wallet"] = "permissions-wallet";
})(PermissionsType = exports.PermissionsType || (exports.PermissionsType = {}));
var SignerMessageKind;
(function (SignerMessageKind) {
    SignerMessageKind["client"] = "signer-client";
    SignerMessageKind["ready"] = "signer-ready";
    SignerMessageKind["success"] = "signer-client-success";
    SignerMessageKind["fail"] = "signer-client-failure";
})(SignerMessageKind = exports.SignerMessageKind || (exports.SignerMessageKind = {}));
// Transaction Types
var TransactionMessageKind;
(function (TransactionMessageKind) {
    TransactionMessageKind["client"] = "transaction-client";
    TransactionMessageKind["ready"] = "transaction-ready";
    TransactionMessageKind["success"] = "transaction-client-success";
    TransactionMessageKind["fail"] = "transaction-client-failure";
})(TransactionMessageKind = exports.TransactionMessageKind || (exports.TransactionMessageKind = {}));
