// @astrox/connection/lib/cjs/connections/nnsConnection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NNSConnection = void 0;
const baseConnection_1 = require("./baseConnection");
const nns_dapp_idl_1 = __importDefault(require("../canisters/nns-dapp.idl"));
const nns_dapp_cert_idl_1 = __importDefault(require("../canisters/nns-dapp-cert.idl"));
const constants_1 = require("../utils/constants");
class NNSConnection extends baseConnection_1.BaseConnection {
    constructor(identity, delegationIdentity, actor, agent, nnsCanisterId) {
        super(identity, delegationIdentity, nnsCanisterId !== null && nnsCanisterId !== void 0 ? nnsCanisterId : constants_1.NNS_CANISTER_ID, nns_dapp_idl_1.default, actor, agent);
        this.identity = identity;
        this.delegationIdentity = delegationIdentity;
        this.actor = actor;
        this.agent = agent;
    }
    get accountDetails() {
        return this._accountDetails;
    }
    /**
     * create connection
     * @param identity
     * @param delegationIdentity
     * @param actor
     * @param agent
     * @function createConnection
     * @returns {NNSConnection}
     */
    static createConnection(identity, delegationIdentity, actor, agent) {
        return new NNSConnection(identity, delegationIdentity, actor, agent);
    }
    /**
     * create Actor with DelegationIdentity
     * @param delegationIdentity
     * @param nnsCanisterId
     * @function {function name}
     * @returns {type} {description}
     */
    static async createActor(delegationIdentity, nnsCanisterId) {
        const actor = await baseConnection_1._createActor(nns_dapp_idl_1.default, nnsCanisterId !== null && nnsCanisterId !== void 0 ? nnsCanisterId : constants_1.NNS_CANISTER_ID, delegationIdentity);
        return actor;
    }
    static async getTransactions({ nnsActor, delegationIdentity, }, { page_size, offset, account_identifier, }) {
        const actor = nnsActor !== null && nnsActor !== void 0 ? nnsActor : (await NNSConnection.createActor(delegationIdentity)).actor;
        const result = await actor.get_transactions({
            page_size,
            offset,
            account_identifier,
        });
        return result;
    }
    /**
     * get NNS Actor, used internally
     * @param nnsCanisterId
     * @function {function name}
     * @returns {type} {description}
     */
    async getNNSActor(nnsCanisterId) {
        const actor = await this._getActor(nnsCanisterId !== null && nnsCanisterId !== void 0 ? nnsCanisterId : constants_1.NNS_CANISTER_ID, nns_dapp_idl_1.default);
        return actor;
    }
    /**
     * get NNS Actor, used internally
     * @param nnsCanisterId
     * @function {function name}
     * @returns {type} {description}
     */
    async getNNSActorCert(nnsCanisterId) {
        const actor = await this._getActor(nnsCanisterId !== null && nnsCanisterId !== void 0 ? nnsCanisterId : constants_1.NNS_CANISTER_ID, nns_dapp_cert_idl_1.default);
        return actor;
    }
    /**
     * when NNSConnection is created, we can get account created to NNS.
     * Even we can just calculate the login principal to NNS DApp, however,
     * The NNS DApp stores and create account, thus, a new Identity login will get NO ACCOUNT created by default.
     * We need to manually create account using `add_account` when no account found.
     *
     * @param cert
     * @function {function name}
     * @returns {type} {description}
     */
    async getAccount(cert) {
        const actor = cert === true ? await this.getNNSActorCert() : await this.getNNSActor();
        const response = await baseConnection_1.executeWithLogging(() => actor.get_account());
        if (response === { AccountNotFound: null }) {
            return undefined;
        }
        else {
            this._accountDetails = response['Ok'];
            return response['Ok'];
        }
    }
    /**
     * create account when new identity logined to NNS
     * @function {function name}
     * @returns {type} {description}
     */
    async addAccount() {
        const actor = await this.getNNSActor();
        const response = await baseConnection_1.executeWithLogging(() => actor.add_account());
        return response;
    }
}
exports.NNSConnection = NNSConnection;
// export const requestNNSDelegation = async (
//   identity: SignIdentity,
// ): Promise<DelegationIdentity> => {
//   const tenMinutesInMsec = 10 * 1000 * 60;
//   const date = new Date(Date.now() + tenMinutesInMsec);
//   return requestDelegation(identity, { canisterId, date });
// };
