// @astrox/connection/lib/cjs/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ADDRESS_TYPES = exports.CANISTER_MAX_LENGTH = exports.CANISTER_REGEX = exports.ALPHANUM_REGEX = exports.PRINCIPAL_REGEX = exports.MAX_TRANSACTION_DECISION_MILSECONDS = exports.IC_ROCKS = exports.IC_EXPLORER = exports.ROSETTA_URL = exports.NET_ID = exports.TRANSACTION_FEE = exports.TOP_UP_CANISTER_MEMO = exports.CREATE_CANISTER_MEMO = exports.SUB_ACCOUNT_BYTE_LENGTH = exports.NNS_CANISTER_ID = exports.LEDGER_CANISTER_ID = void 0;
exports.LEDGER_CANISTER_ID = 'ryjl3-tyaaa-aaaaa-aaaba-cai';
exports.NNS_CANISTER_ID = 'qoctq-giaaa-aaaaa-aaaea-cai';
exports.SUB_ACCOUNT_BYTE_LENGTH = 32;
exports.CREATE_CANISTER_MEMO = BigInt(0x41455243); // CREA,
exports.TOP_UP_CANISTER_MEMO = BigInt(0x50555054); // TPUP
exports.TRANSACTION_FEE = BigInt(10000);
exports.NET_ID = {
    blockchain: 'Internet Computer',
    network: '00000000000000020101',
};
exports.ROSETTA_URL = 'https://rosetta-api.internetcomputer.org';
exports.IC_EXPLORER = 'https://dashboard.internetcomputer.org';
exports.IC_ROCKS = 'https://ic.rocks';
exports.MAX_TRANSACTION_DECISION_MILSECONDS = 120000;
exports.PRINCIPAL_REGEX = /(\w{5}-){10}\w{3}/;
exports.ALPHANUM_REGEX = /^[a-zA-Z0-9]+$/;
exports.CANISTER_REGEX = /(\w{5}-){4}\w{3}/;
exports.CANISTER_MAX_LENGTH = 27;
exports.ADDRESS_TYPES = {
    PRINCIPAL: 'principal',
    ACCOUNT: 'accountId',
    CANISTER: 'canister',
    ERC20: 'erc20',
    UNKNOWN: 'unknown',
};
